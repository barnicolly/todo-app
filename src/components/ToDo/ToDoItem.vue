<template>
  <div>
    <input
      class="todo-item__checkbox"
      type="checkbox"
      :id="id"
      :checked="isCompleted"
      @change="updateDoneStatus(id)" />

    <div v-if="!isEditing" class="todo-item__content">
      <label class="todo-item__label" @click="toggleToItemEditForm">{{
        title
      }}</label>
      <div class="todo-item__controls">
        <button class="btn btn--delete" type="button" @click="deleteToDo(id)">
          <font-awesome-icon icon="trash-alt" />
        </button>
      </div>
    </div>
    <EditForm
      class="todo-item__content"
      v-else
      :id="id"
      :title="title"
      @item-edited="itemEdited"
      @edit-cancelled="editCancelled"></EditForm>
  </div>
</template>

<script setup lang="ts">
import { useTodoStore } from '@/store/todo';
import { computed, ref } from 'vue';
import { type Todo } from '@/types/todo';
import EditForm from './EditForm.vue';

const props = defineProps<Todo>();

const isCompleted = computed(() => props.completed);
const isEditing = ref(false);

const store = useTodoStore();
const { deleteToDo, updateDoneStatus } = store;

function toggleToItemEditForm() {
  isEditing.value = true;
}

function editCancelled() {
  isEditing.value = false;
}

function itemEdited(newTitle: string) {
  store.editToDo({ title: newTitle, toDoId: props.id });
  isEditing.value = false;
}
</script>
