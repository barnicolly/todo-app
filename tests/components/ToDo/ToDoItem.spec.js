import { shallowMount, mount } from '@vue/test-utils';
import { faker } from '@faker-js/faker';
import { Store } from 'vuex';
import ToDoItem from '@/components/ToDo/ToDoItem.vue';

// todo-misha переделать селекторы как будут классы;
describe('ToDoItem.vue', () => {
  let wrapper;
  const id = faker.string.uuid();
  const title = faker.lorem.sentence();
  const completed = false;

  const deleteToDoMock = jest.fn();
  const editToDoMock = jest.fn();
  const updateDoneStatusMock = jest.fn();
  const $store = new Store({
    mutations: {
      deleteToDo: deleteToDoMock,
      editToDo: editToDoMock,
      updateDoneStatus: updateDoneStatusMock,
    },
  });

  beforeEach(() => {
    wrapper = shallowMount(ToDoItem, {
      props: {
        title,
        id,
        completed,
      },
      global: {
        mocks: {
          $store,
        },
      },
    });
  });

  it('should correct initialize, checked false', () => {
    const input = wrapper.find('input');
    expect(wrapper.vm.isEditing).toBeFalsy();
    expect(input.element.checked).toBeFalsy();
  });

  it('should correct initialize, checked true', () => {
    wrapper = shallowMount(ToDoItem, {
      props: {
        title,
        id,
        completed: true,
      },
    });
    const input = wrapper.find('input');
    expect(wrapper.vm.isEditing).toBeFalsy();
    expect(input.element.checked).toBeTruthy();
  });

  it('click edit btn', () => {
    wrapper.find('button:nth-child(1)').trigger('click');
    expect(wrapper.vm.isEditing).toBeTruthy();
  });

  it('click edit btn -> save', async () => {
    await wrapper.find('button:nth-child(1)').trigger('click');
    expect(wrapper.vm.isEditing).toBeTruthy();

    wrapper.vm.itemEdited('changed-title');
    expect(editToDoMock.mock.calls.length).toBe(1);
    expect(editToDoMock.mock.calls[0][1]).toStrictEqual({ title: 'changed-title', toDoId: id });
  });

  it('click checkbox, init checked false', async () => {
    const checkbox = wrapper.find('input');
    await checkbox.setChecked();
    expect(checkbox.element.checked).toBeTruthy();
    expect(updateDoneStatusMock.mock.calls.length).toBe(1);
    expect(updateDoneStatusMock.mock.calls[0][1]).toBe(id);
  });

  it('click checkbox, init checked true', async () => {
    wrapper = shallowMount(ToDoItem, {
      props: {
        title,
        id,
        completed: true,
      },
      global: {
        mocks: {
          $store,
        },
      },
    });
    const checkbox = wrapper.find('input');
    checkbox.element.checked = false;
    expect(updateDoneStatusMock.mock.calls.length).toBe(1);
    expect(updateDoneStatusMock.mock.calls[0][1]).toBe(id);
  });

  it('click delete btn', () => {
    wrapper.find('button:nth-child(2)').trigger('click');
    expect(deleteToDoMock.mock.calls.length).toBe(1);
    expect(deleteToDoMock.mock.calls[0][1]).toBe(id);
  });
});

describe('ToDoItem.vue Snapshot', () => {
  it('has same HTML structure, without edit', () => {
    const wrapper = mount(ToDoItem, {
      props: {
        title: 'some-title',
        id: 'some-uuid',
        completed: false,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('has same HTML structure, edit', async () => {
    const wrapper = mount(ToDoItem, {
      props: {
        title: 'some-title',
        id: 'some-uuid',
        completed: false,
      },
    });
    await wrapper.find('button:nth-child(1)').trigger('click');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
