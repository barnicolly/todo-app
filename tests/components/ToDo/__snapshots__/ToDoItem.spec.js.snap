// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoItem.vue Snapshot has same HTML structure, edit 1`] = `
<form>
  <div><input id="some-uuid" type="text" autocomplete="off"></div>
  <div><button type="button">Сохранить</button><button type="button">Отмена</button></div>
</form>
`;

exports[`ToDoItem.vue Snapshot has same HTML structure, without edit 1`] = `
<div>
  <div><input type="checkbox" id="some-uuid"><label for="some-uuid">some-title</label></div>
  <div><button type="button">Редактировать</button><button type="button">Удалить</button></div>
</div>
`;
